sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala

```diff
@@ -245,6 +246,28 @@ object DataSourceV2Strategy extends Strategy with PredicateHelper {
       }.toArray
       DeleteFromTableExec(r.table.asDeletable, filters) :: Nil
 
+    case UpdateTable(r: DataSourceV2Relation, attrs, values, condition) =>
+      val nested = attrs.asInstanceOf[Seq[Any]].filterNot(_.isInstanceOf[AttributeReference])
+      if (nested.nonEmpty) {
+        throw new AnalysisException(s"Update only support non-nested fields. Nested: $nested")
+      }
+
+      val attrsNames = attrs.map(_.name)
+      // fail if any updated value cannot be converted.
+      val updatedValues = values.map {
+        v => DataSourceStrategy.translateExpression(v).getOrElse(
+          throw new AnalysisException(s"Exec update failed:" +
+              s" cannot translate update set to source expression: $v"))
+      }
+      // fail if any filter cannot be converted. correctness depends on removing all matching data.
+      val filters = condition.map(
+        splitConjunctivePredicates(_).map {
+          f => DataSourceStrategy.translateFilter(f).getOrElse(
+            throw new AnalysisException(s"Exec update failed:" +
+                s" cannot translate expression to source filter: $f"))
+        }.toArray).getOrElse(Array.empty[Filter])
+      UpdateTableExec(r.table.asUpdatable, attrsNames, updatedValues, filters)::Nil
```

Comment:
We should normalize the attribute name. Let's say a table has a column `id`, and users write `UPDATE ... ID=1`, then the `AttributeReference` will be named `ID` instead of `id`.

The same thing should also apply to filters. You can take a look at `DataSourceStrategy.normalizeFilters` and see where we call it.

Reply:
Fixed.